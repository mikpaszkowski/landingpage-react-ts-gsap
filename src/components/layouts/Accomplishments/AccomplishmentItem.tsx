import * as React from "react";
import styled from "styled-components";
import { fontTypes } from "../../../styles/constants";
import { device } from "../../../styles/responsive";

const AccomplishmentItemWrapper = styled.div``;

const Headline = styled.h1`
  display: block;
  font-family: ${fontTypes.primaryFont};
  font-style: normal;
  font-weight: normal;
  font-size: 8rem;
  line-height: 117%;
  margin-bottom: 1.6rem;
  text-align: center;
  font-feature-settings: "haln" on;
  color: ${({ theme }) => theme.palette.common.theme};

  @media ${device.mobileMedium} {
    margin-bottom: 0.6rem;
  }
`;

const Description = styled.p`
  display: block;
  font-family: ${fontTypes.secondaryFont};
  font-style: normal;
  font-weight: normal;
  font-size: 2.4rem;
  line-height: 3.6rem;
  margin: 0 4rem;
  text-align: center;
  color: ${({ theme }) => theme.palette.primary.font};
`;

interface AccomplishmentProps {
  headline: string;
  description: string;
}

export const AccomplishmentItem = ({
  headline,
  description,
}: AccomplishmentProps) => {
  return (
    <AccomplishmentItemWrapper>
      <Headline>{headline}</Headline>
      <Description>{description}</Description>
    </AccomplishmentItemWrapper>
  );
};
