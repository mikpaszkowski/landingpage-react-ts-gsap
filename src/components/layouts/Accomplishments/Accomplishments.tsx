import * as React from "react";

import styled from "styled-components";
import { RecommendationItem } from "./AccomplishmentItem";
import { device } from "../../../styles/responsive";
import gsap from "gsap";
import { ScrollTrigger } from "gsap/ScrollTrigger";

gsap.registerPlugin(ScrollTrigger);

const RecommendationWrapper = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 11.1rem;
  margin: 22rem 21.9rem;
  will-change: transform;

  @media ${device.mobileMedium} {
    display: block;
    margin: 6.4rem 0;
    padding: 0 2.4rem;
  }
`;

export const Recommendation = () => {
  const { useRef, useEffect } = React;

  const sectionRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    ScrollTrigger.matchMedia({
      "(min-width: 1440px)": function () {
        gsap.fromTo(
          sectionRef.current,
          {
            opacity: 0,
            y: "+=100",
          },
          {
            duration: 0.7,
            delay: 0.3,
            opacity: 1,
            y: 0,
            ease: "power4.out",
            scrollTrigger: {
              trigger: `${sectionRef.current?.className}`,
              start: "top 20%",
              markers: true,
            },
          }
        );
      },

      "(max-width: 375px)": function () {
        gsap.fromTo(
          sectionRef.current,
          {
            opacity: 0,
            y: "-=60",
          },
          {
            duration: 0.5,
            delay: 0.3,
            opacity: 1,
            y: 0,
            ease: "none",
            scrollTrigger: {
              trigger: `${sectionRef.current?.className}`,
              start: "top+=500 20%",
              markers: true,
            },
          }
        );
      },
    });
  });

  return (
    <RecommendationWrapper ref={sectionRef}>
      <RecommendationItem
        headline="230+"
        description="Scientifically based tests"
      />
      <RecommendationItem
        headline="10,000+"
        description="Medical recommendations"
      />
      <RecommendationItem headline="25M+" description="Students learning" />
    </RecommendationWrapper>
  );
};
