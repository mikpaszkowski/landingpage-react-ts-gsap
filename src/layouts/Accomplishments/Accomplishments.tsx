import * as React from "react";

import styled from "styled-components";
import { AccomplishmentItem } from "./AccomplishmentItem";
import { accomplishmentItems } from "../../data/staticContents/accomplishments/accomplishmentItemList";
import { device } from "../../styles/responsive";
import gsap from "gsap";
import { ScrollTrigger } from "gsap/dist/ScrollTrigger";
import {
  laptopLarge,
  mobileMedium,
  tabletLarge,
} from "../../animations/accomplishments";

gsap.registerPlugin(ScrollTrigger);

const AccomplishmentsWrapper = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 11.1rem;
  margin: 22rem 21.9rem;
  will-change: transform;

  @media ${device.laptopSmall} {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: center;
    align-content: center;
    margin: 22rem 21.9rem;
  }

  @media ${device.tabletLarge} {
    margin: 15rem 0;
    flex-direction: column;
  }

  @media ${device.mobileLarge} {
    margin: 6.4rem 0;
    flex-direction: column;
  }
`;

export const Accomplishments = () => {
  const { useRef, useEffect } = React;

  const sectionRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    ScrollTrigger.matchMedia({
      "(min-width: 1440px)": () => laptopLarge(sectionRef),
      "(min-width: 500px)": () => tabletLarge(sectionRef),
      "(max-width: 499px)": () => mobileMedium(sectionRef),
    });
  });

  return (
    <AccomplishmentsWrapper ref={sectionRef}>
      {accomplishmentItems.map(({ id, ...props }) => (
        <AccomplishmentItem key={id} {...props} />
      ))}
    </AccomplishmentsWrapper>
  );
};
