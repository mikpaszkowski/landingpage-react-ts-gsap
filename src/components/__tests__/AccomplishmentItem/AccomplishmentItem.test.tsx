import { cleanup } from "@testing-library/react";
import "jest-styled-components";
import {
  renderWithLightTheme,
  mountWithLightTheme,
  mountWithDarkTheme,
} from "../helpers";
import { AccomplishmentItem } from "../../../layouts/Accomplishments/AccomplishmentItem";

afterEach(() => {
  cleanup();
});

it("check render correctly <AccomplishmentItem /> component", () => {
  const props = {
    headline: "10,000+",
    description: "Medical recommendations",
  };
  const tree = renderWithLightTheme(<AccomplishmentItem {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("check AccomplishmentItem props", () => {
  test("should have headline prop displayed", () => {
    const props = {
      headline: "10,000+",
      description: "Medical recommendations",
    };
    const wrapper = mountWithLightTheme(<AccomplishmentItem {...props} />).find(
      AccomplishmentItem
    );
    expect(wrapper.find("h1").text()).toEqual(props.headline);
  });

  test("should have description prop displayed", () => {
    const props = {
      headline: "10,000+",
      description: "Medical recommendations",
    };
    const wrapper = mountWithLightTheme(<AccomplishmentItem {...props} />).find(
      AccomplishmentItem
    );
    expect(wrapper.find("p").text()).toEqual(props.description);
  });
});

describe("check AccomplishmentItem props data type", () => {
  test("should headline prop be string type", () => {
    const props = {
      headline: "10,000+",
      description: "Medical recommendations",
    };
    const wrapper = mountWithLightTheme(<AccomplishmentItem {...props} />).find(
      AccomplishmentItem
    );
    expect(typeof wrapper.prop("headline")).toEqual("string");
  });

  test("should description prop be string type", () => {
    const props = {
      headline: "10,000+",
      description: "Medical recommendations",
    };
    const wrapper = mountWithLightTheme(<AccomplishmentItem {...props} />).find(
      AccomplishmentItem
    );
    expect(typeof wrapper.prop("description")).toEqual("string");
  });
});

describe("check AccomplishmentItem css properties", () => {
  test("should have headline color: #FABD20", () => {
    const props = {
      headline: "10,000+",
      description: "Medical recommendations",
    };
    const wrapper = mountWithLightTheme(<AccomplishmentItem {...props} />).find(
      AccomplishmentItem
    );
    expect(wrapper.find("h1")).toHaveStyleRule("color", "#FABD20");
  });

  describe("check description font color when", () => {
    test("theme set to dark - #FFFFFF", () => {
      const props = {
        headline: "10,000+",
        description: "Medical recommendations",
      };
      const wrapper = mountWithDarkTheme(
        <AccomplishmentItem {...props} />
      ).find(AccomplishmentItem);
      expect(wrapper.find("p")).toHaveStyleRule("color", "#FFFFFF");
    });

    test("theme set to light - p color #131E07", () => {
      const props = {
        headline: "10,000+",
        description: "Medical recommendations",
      };
      const wrapper = mountWithLightTheme(
        <AccomplishmentItem {...props} />
      ).find(AccomplishmentItem);
      expect(wrapper.find("p")).toHaveStyleRule("color", "#131E07");
    });

    test("theme set to dark - p color #FFFFFF", () => {
      const props = {
        headline: "10,000+",
        description: "Medical recommendations",
      };
      const wrapper = mountWithDarkTheme(
        <AccomplishmentItem {...props} />
      ).find(AccomplishmentItem);
      expect(wrapper.find("p")).toHaveStyleRule("color", "#FFFFFF");
    });
  });
});
