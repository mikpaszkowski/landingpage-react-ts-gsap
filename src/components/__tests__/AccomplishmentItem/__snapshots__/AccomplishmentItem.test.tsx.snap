// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check render correctly <AccomplishmentItem /> component 1`] = `
.c0 {
  display: block;
  font-family: 'Rozha One',serif;
  font-style: normal;
  font-weight: normal;
  font-size: 8rem;
  line-height: 117%;
  margin-bottom: 1.6rem;
  text-align: center;
  font-feature-settings: "haln" on;
  color: #FABD20;
}

.c1 {
  display: block;
  font-family: 'Reem Kufi',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 2.4rem;
  line-height: 3.6rem;
  margin: 0 4rem;
  text-align: center;
  color: #131E07;
}

@media (max-width:375px) {
  .c0 {
    margin-bottom: 0.6rem;
  }
}

<div
  className=""
>
  <h1
    className="c0"
  >
    10,000+
  </h1>
  <p
    className="c1"
  >
    Medical recommendations
  </p>
</div>
`;
